DSL/conditions.md

```diff
@@ -0,0 +1,237 @@
+# The Condition DSL
+
+This document provides an overview of a declarative Domain Specific Language (DSL) designed to check various conditions for upgrade and downgrade rules in this repository.
+
+## Why Not Use JSON Schema Validation?
+
+There are several reasons for opting against JSON Schema as the schema language for these conditions:
+
+- **Simplicity**: Implementers do not need a full-blown JSON Schema implementation to support upgrade/downgrade rules in their language of choice. This reduces the complexity and learning curve for developers.
+- **Performance**: JSON Schema validation can be slow, especially on Alterschema. By focusing on a subset of operations, we achieve much better performance.
+- **Specific Needs**: The modifications needed are typically only 1-2 levels deep, and only 20-30% of JSON Schema constraints are necessary to satisfy the conditions for upgrades/downgrades.
+
+## DSL Structure
+
+The DSL is structured as an array of conditions, with each condition containing three mandatory fields: `path`, `operation`, and `value`.
+
+### Condition Fields
+
+- `path`: Specifies the path to the target in the schema where the operation will be performed. It takes an array. Each element of the array can be:
+  - A string
+  - An object which can have pre-defined properties. Which is: `pattern`.
+    Given a pattern, all the keywords matching the pattern undergo condition check and transformation iteratively.  
+    Example: 
+
+    #### Schema
+    ```json
+    {    
+      "$schema": "https://json-schema.org/draft-06/schema",
+      "properties": {
+        "foo": {
+            "required": true
+        },
+        "bar": {
+            "required": false
+        }
+      }
+    }
+    ```
+    #### Condition
+    ```json
+    [
+      { "path": [ "properties", { "pattern": ".*" } ], "operation": "has-key", "value": "required" }
```

Comment:
There is a situation where we should use `enum` instead of `pattern`, still I think that's good as it works for both and even more key/string matching logics...

Reply:
Can you show me a concrete example for that enum case?

Reply:
@jviotti for the unknown keywords, I thought of having enum.But thats resolved. Can you use this even for array items??
I think just using this: {} is much better